sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Main.java

```diff
@@ -74,30 +78,97 @@
  * <p>See <a href="http://datalab.cs.pdx.edu/niagaraST/NEXMark/">
  * http://datalab.cs.pdx.edu/niagaraST/NEXMark/</a>
  */
-public class Main<OptionT extends NexmarkOptions> {
+public class Main {
+
+  private static class Run implements Runnable {
+    final NexmarkLauncher<NexmarkOptions> nexmarkLauncher;
+    final NexmarkConfiguration configuration;
+    final BlockingQueue results;
+    NexmarkPerf perf = null;
+    Exception exception = null;
+
+    private Run(String[] args, NexmarkConfiguration configuration, BlockingQueue results) {
+      NexmarkOptions options = PipelineOptionsFactory.fromArgs(args).as(NexmarkOptions.class);
+      this.nexmarkLauncher = new NexmarkLauncher<>(options);
+      this.configuration = configuration;
+      this.results = results;
+    }
+
+    @Override
+    public void run() {
+      try {
+        try {
+          perf = nexmarkLauncher.run(configuration);
+        } catch (IOException | RuntimeException e) {
+          exception = e;
+        } finally {
+          results.put(this);
+        }
+      } catch (InterruptedException ie) {
+        return;
+      }
+    }
+
+    private NexmarkPerf get() throws IOException {
+      if (exception != null) {
+        if (exception instanceof IOException) {
+          throw (IOException) exception;
+        } else if (exception instanceof RuntimeException) {
+          throw (RuntimeException) exception;
+        }
+        // This should be unreachable.
+      }
+      assert perf != null;
+      return perf;
+    }
+  }
 
   /** Entry point. */
-  void runAll(OptionT options, NexmarkLauncher nexmarkLauncher) throws IOException {
+  void runAll(String[] args) throws IOException {
     Instant start = Instant.now();
+    NexmarkOptions options =
+        PipelineOptionsFactory.fromArgs(args).withValidation().as(NexmarkOptions.class);
+    ExecutorService executor = Executors.newFixedThreadPool(options.getNexmarkParallel());
+    BlockingQueue<Run> results = new LinkedBlockingQueue();
     Map<NexmarkConfiguration, NexmarkPerf> baseline = loadBaseline(options.getBaselineFilename());
     Map<NexmarkConfiguration, NexmarkPerf> actual = new LinkedHashMap<>();
     Iterable<NexmarkConfiguration> configurations = options.getSuite().getConfigurations(options);
 
     boolean successful = true;
     try {
-      // Run all the configurations.
+      // Schedule all the configurations.
       for (NexmarkConfiguration configuration : configurations) {
-        NexmarkPerf perf = nexmarkLauncher.run(configuration);
-        if (perf != null) {
-          if (perf.errors == null || perf.errors.size() > 0) {
-            successful = false;
+        executor.execute(new Run(args, configuration, results));
+        if (options.getNexmarkParallel() > 1) {
+          // Stagger start
+          try {
+            Thread.sleep(2000);
```

Comment:
I heard we were hitting quota issues on spinup. Happy to drop it if @Ardagan thinks it is unneeded.

Reply:
This doesn't really address quota issue directly. I believe the problem was with running too many post-commit tests in parallel and I've addressed that issue in https://github.com/apache/beam/pull/6125.

What we can do instead is add retry in case of quota issue. But again, that is not a panacea and does not address issue directly.

My opinion is that we should remove this check as of now.